Reorganize and update the Java documentation in the README file in a … (#166)
bazelbuild/rules_fuzzing: README.md

The crash is saved under `/tmp/fuzzing/artifacts` and can be further inspected.

### Java fuzzing

> You can write `java_fuzz_tests` through the [Jazzer][jazzer-doc] fuzzing engine. You will need to enable it in your WORKSPACE `rules_fuzzing_dependencies` call:

Author: Done.

```python

> rules_fuzzing_dependencies(jazzer = True)

Reviewer: It may be clearer to give the full working snippet here (excluding the `http_archive`), even if it repeats a few more lines. Otherwise the `rules_fuzzing_init()` call is a bit too easy to miss:

```python
load("@rules_fuzzing//fuzzing:repositories.bzl", "rules_fuzzing_dependencies")

rules_fuzzing_dependencies(jazzer = True)

load("@rules_fuzzing//fuzzing:init.bzl", "rules_fuzzing_init")

rules_fuzzing_init()

load("@jazzer//:repositories.bzl", "jazzer_dependencies")

jazzer_dependencies()

load("@jazzer//:init.bzl", "jazzer_init")

jazzer_init()
```

Author: Good call, done.

load("@rules_fuzzing//fuzzing:init.bzl", "rules_fuzzing_init")

rules_fuzzing_init()

load("@jazzer//:repositories.bzl", "jazzer_dependencies")

jazzer_dependencies()

load("@jazzer//:init.bzl", "jazzer_init")

jazzer_init()
```

> To use Jazzer, it is convienient to also define a `.bazelrc` configuration, similar to the C++ libFuzzer one above:

Reviewer: ```suggestion
To use Jazzer, it is convenient to also define a `.bazelrc` configuration, similar to the C++ libFuzzer one above:
```

Author: Oops, good catch.

```

> # Force the use of Clang for all builds (needed to build Jazzer).

Reviewer: Would you say that g++ compilability is desirable here? The [most recent Jazzer commit](https://github.com/CodeIntelligenceTesting/jazzer) achieves that and I could update the `repository.bzl` reference. Of course, since `-fsanitize=fuzzer-no-link` is not supported, this would just enable Java-only fuzzing.

Author: > Would you say that g++ compilability is desirable here?

I'm not really sure, as I'm not familiar with the typical toolchains used to build native code mixed with Java code. I guess we can consider adding support for it if it ends up breaking some of our users.

> # Force the use of Clang for all builds (needed to build Jazzer).
> build --action_env=CC=clang-10
> build --action_env=CXX=clang++-10

Reviewer: Fixing a particular version of `clang` may be problematic here (although it is of course the right thing to do in `rules_fuzzing`'s own `.bazelrc`). How about:

```suggestion
# Force the use of Clang for all builds (Jazzer requires at least clang 9).
build --action_env=CC=clang
build --action_env=CXX=clang++
```

Author: Good catch - fixed.

# Define --config=jazzer for Jazzer without sanitizer (Java only).
build:jazzer --@rules_fuzzing//fuzzing:java_engine=@rules_fuzzing//fuzzing/engines:jazzer
build:jazzer --@rules_fuzzing//fuzzing:cc_engine_instrumentation=jazzer
build:jazzer --@rules_fuzzing//fuzzing:cc_engine_sanitizer=none

# Define --config=asan-jazzer for Jazzer + ASAN.
build:asan-jazzer --@rules_fuzzing//fuzzing:java_engine=@rules_fuzzing//fuzzing/engines:jazzer
build:asan-jazzer --@rules_fuzzing//fuzzing:cc_engine_instrumentation=jazzer
build:asan-jazzer --@rules_fuzzing//fuzzing:cc_engine_sanitizer=asan
```

A Java fuzz test is specified using a [`java_fuzz_test` rule](/docs/java-fuzzing-rules.md#java_fuzz_test). In the most basic form, a Java fuzz test consists of a single `.java` file with a class that defines a function `public static fuzzerTestOneInput(byte[] input)`.

> Create the `com/example/JavaFuzzTest.java` file in your workspace root, as follows:

Reviewer: Following the Maven layout simplifies the `java_fuzz_test` target below:

```suggestion
Create the `src/com/example/JavaFuzzTest.java` file in your workspace root, as follows:
```

Author: Done.

}
```

You should now define the corresponding target in the `BUILD` file, which looks very much like a regular `java_binary`:

```python
load("@rules_fuzzing//fuzzing:java_defs.bzl", "java_fuzz_test")

java_fuzz_test(
    name = "JavaFuzzTest",

>     srcs = ["com/example/JavaFuzzTest.java"],
>     # target_class is not needed if using the Maven directory layout.
>     target_class = "com.example.JavaFuzzTest",

Reviewer: With the path change above, this allows to drop `target_class`:
```suggestion
    srcs = ["src/com/example/JavaFuzzTest.java"],
    # target_class is not needed if using the Maven directory layout.
    # target_class = "com.example.JavaFuzzTest",
```

Author: Done.
